runtime/runtime/src/global_contracts.rs

```diff
@@ -228,54 +228,70 @@ fn forward_distribution_next_shard(
     ) -> Result<Option<ContractCode>, StorageError>;
 }
 
-impl AccountContractStoreExt for AccountContract {
+impl AccountContractAccessExt for AccountContract {
     fn code(
-        &self,
-        account_id: &AccountId,
+        self,
+        local_account_id: &AccountId,
         store: &TrieUpdate,
     ) -> Result<Option<ContractCode>, StorageError> {
-        let Some(key) = TrieKey::for_account_contract_code(account_id, self) else {
-            return Ok(None);
+        let local_hash = match GlobalContractIdentifier::try_from(self) {
+            Ok(identifier) => return identifier.code(store),
+            Err(None) => return Ok(None),
+            Err(Some(local_hash)) => local_hash,
         };
-        let code_hash = match self {
-            AccountContract::None | AccountContract::GlobalByAccount(_) => None,
-            AccountContract::Local(hash) | AccountContract::Global(hash) => Some(*hash),
-        };
-        store
-            .get(&key, AccessOptions::DEFAULT)
-            .map(|opt| opt.map(|code| ContractCode::new(code, code_hash)))
+        let key = TrieKey::ContractCode { account_id: local_account_id.clone() };
+        let code = store.get(&key, AccessOptions::DEFAULT)?;
+        Ok(code.map(|code| ContractCode::new(code, Some(local_hash))))
     }
 
-    fn hash(&self, store: &TrieUpdate) -> Result<CryptoHash, StorageError> {
-        let hash = match self {
-            AccountContract::None => CryptoHash::default(),
-            AccountContract::Local(code_hash) | AccountContract::Global(code_hash) => *code_hash,
-            AccountContract::GlobalByAccount(account_id) => {
-                let identifier = GlobalContractIdentifier::AccountId(account_id.clone());
-                let key = TrieKey::GlobalContractCode { identifier: identifier.into() };
-                let value_ref = store
-                    .get_ref(&key, KeyLookupMode::MemOrFlatOrTrie, AccessOptions::DEFAULT)?
-                    .ok_or_else(|| {
-                        let TrieKey::GlobalContractCode { identifier } = key else {
-                            unreachable!()
-                        };
-                        StorageError::StorageInconsistentState(format!(
-                            "Global contract identifier not found {:?}",
-                            identifier
-                        ))
-                    })?;
-                value_ref.value_hash()
-            }
+    fn hash(self, store: &TrieUpdate) -> Result<CryptoHash, StorageError> {
+        match GlobalContractIdentifier::try_from(self) {
+            Ok(gci) => return gci.hash(store),
+            Err(None) => return Ok(CryptoHash::default()),
+            Err(Some(local_hash)) => Ok(local_hash),
+        }
+    }
+}
+
+pub(crate) trait GlobalContractAccessExt {
+    fn hash(self, store: &TrieUpdate) -> Result<CryptoHash, StorageError>;
+    fn code(self, store: &TrieUpdate) -> Result<Option<ContractCode>, StorageError>;
```

Comment:
It's baffling to me that None can be returned but it's outside of your PR so no need to change it. If you know why though please share :)